Parse error changes (and other random stuff)

patrick-east opened the pull request:
<!--

Thanks for submitting a PR to OPA!

Before pressing 'Create pull request' please read the checklist below.

* All code changes should be accompanied with tests. If you are not
modifying any tests, just provide a short explanation of why updates
to tests are not necessary. In addition to helping catch bugs, tests
are extremely helpful in providing _context_ that explains how your
changes can be used.

* All changes to public APIs **must** be accompanied with
docs. Examples of public APIs include built-in functions,
config fields, and of course, exported Go types/functions/constants/etc.

* Commit messages should explain _why_ you made the changes, not what
you changed. Use active voice. Keep the subject line under 50
characters or so.

* All commits must be signed off by the author. If you are not
familiar with signing off, see CONTRIBUTING.md below.

For more information on contributing to OPA see:

* [CONTRIBUTING.md](https://github.com/open-policy-agent/opa/blob/master/CONTRIBUTING.md)
  for high-level contribution guidelines.

* [DEVELOPMENT.md](https://github.com/open-policy-agent/opa/blob/master/docs/devel/DEVELOPMENT.md)
  for development workflow and environment setup.

-->

patrick-east committed:
Parse error improvements (mostly for parseTerm)

Signed-off-by: Patrick East <east.patrick@gmail.com>

patrick-east committed:
Don't add extra error when failing to parse a rule

Signed-off-by: Patrick East <east.patrick@gmail.com>

patrick-east committed:
More error message improments

Signed-off-by: Patrick East <east.patrick@gmail.com>

patrick-east committed:
Store offset with locations, use it for details about the location

Signed-off-by: Patrick East <east.patrick@gmail.com>

patrick-east committed:
Add explicit error for empty bodies (like the old parser had)

Signed-off-by: Patrick East <east.patrick@gmail.com>

patrick-east committed:
Remove rego.peg

Signed-off-by: Patrick East <east.patrick@gmail.com>

patrick-east commented:
These should take care of any of the remaining error differences/weirdness beyond places where we still have >1 error for the same spot (but they make sense and IMO unless we figure out how to combine them in an elegant way we should just continue sending back >1)

After this the only things left on my todo list are to rename `parser_new.go` and document anything we need to call out in the changelog (I need to go file bugs for a handful of issues so we can reference them in the changelog entry for the debatably backwards incompatible changes/bug fixes).

tsandall commented on ast/parser_new.go:
```diff
@@ -1385,6 +1385,7 @@ func (p *Parser) doScan(skipws bool) {
 
 		p.s.loc.Row = p.s.pos.Row
 		p.s.loc.Col = p.s.pos.Col
+		p.s.loc.Offset = p.s.pos.Offset
```
Do we have to store the position on the parser anymore?

patrick-east replied:
We still use `p.s.pos.End` in a few spots, but I can change to only keep around the end + location rather than the full position (save us a few int's per save 😄)

tsandall replied:
Let's not worry about that for now until we do some more performance analysis.

patrick-east replied:
Was a quick `sed` sorta change, and gave up to* ~5% improvement on memory usage so I figured why not. The most important one IMO being on:

```
ParseBasicABACModule-8                39.0kB ± 0%    37.2kB ± 0%   -4.67%  (p=0.000 n=10+8)
```

tsandall commented on ast/location/location.go:
```diff
@@ -10,10 +10,11 @@ import (
+	File   string `json:"file"` // The name of the source file (which may be empty).
+	Row    int    `json:"row"`  // The line in the source.
+	Col    int    `json:"col"`  // The column in the row.
+	Offset int    `json:"-"`    // The byte offset for the location in the source.
```
I'm assuming we omit this field from the serialized version for backwards compatiblity in test cases? A comment about that would be helpful.

patrick-east replied:
Was mostly to avoid adding it to the like REST API and CLI responses, if I could have (without more intrusive change) I would have not added it to the public Golang API either.

If you think it would be useful to have in the JSON format I'm game to add it. Just didn't want to commit us to exposing it without any particular external use-case.

tsandall replied:
No, that's fine. I just wanted to understand the reason for not exposing it for now. Thanks!

tsandall reviewed (commented):
it's so close...I can taste it...🚀

patrick-east committed:
Test fixups for empty rule body error

Signed-off-by: Patrick East <east.patrick@gmail.com>

patrick-east committed:
Switch to using the location offset everywhere

Signed-off-by: Patrick East <east.patrick@gmail.com>

patrick-east committed:
Only keep lastEnd and tokEnd instead of last and pos on scanner

Not a crazy improvement, but 5% less memory is pretty nice:

```
name                                old time/op    new time/op    delta
ParseModuleRulesBase/1-8              4.51µs ± 1%    4.40µs ± 3%   -2.38%  (p=0.001 n=9+10)
ParseModuleRulesBase/10-8             31.1µs ± 4%    30.4µs ± 3%   -2.16%  (p=0.019 n=10+10)
ParseModuleRulesBase/100-8             298µs ± 1%     304µs ± 6%     ~     (p=1.000 n=10+10)
ParseModuleRulesBase/1000-8           3.35ms ± 8%    3.31ms ± 6%     ~     (p=0.182 n=9+10)
ParseStatementBasicCall-8             3.32µs ± 4%    3.16µs ± 3%   -4.99%  (p=0.000 n=8+10)
ParseStatementMixedJSON-8             84.2µs ± 8%    80.3µs ± 4%   -4.55%  (p=0.005 n=10+10)
ParseStatementSimpleArray/1-8         2.87µs ±22%    2.64µs ± 3%   -7.88%  (p=0.008 n=9+10)
ParseStatementSimpleArray/10-8        9.26µs ±32%    7.80µs ± 3%  -15.72%  (p=0.000 n=10+10)
ParseStatementSimpleArray/100-8       60.7µs ± 6%    60.4µs ± 3%     ~     (p=1.000 n=10+10)
ParseStatementSimpleArray/1000-8       568µs ± 1%     570µs ± 3%     ~     (p=0.796 n=10+10)
ParseStatementNestedObjects/1x1-8     3.49µs ± 0%    3.52µs ± 6%     ~     (p=0.326 n=9+10)
ParseStatementNestedObjects/5x1-8     8.41µs ± 1%    8.58µs ± 6%     ~     (p=0.182 n=9+10)
ParseStatementNestedObjects/10x1-8    14.4µs ± 0%    14.4µs ± 1%     ~     (p=0.712 n=8+10)
ParseStatementNestedObjects/1x5-8     8.18µs ± 0%    8.02µs ± 0%   -1.92%  (p=0.000 n=9+9)
ParseStatementNestedObjects/1x10-8    14.0µs ± 1%    14.1µs ± 5%     ~     (p=0.143 n=10+10)
ParseStatementNestedObjects/5x5-8     5.08ms ± 0%    5.06ms ± 6%     ~     (p=0.156 n=9+10)
ParseBasicABACModule-8                43.5µs ± 2%    42.5µs ± 2%   -2.26%  (p=0.006 n=10+9)

name                                old alloc/op   new alloc/op   delta
ParseModuleRulesBase/1-8              5.90kB ± 0%    5.66kB ± 0%   -4.07%  (p=0.000 n=10+10)
ParseModuleRulesBase/10-8             30.6kB ± 0%    29.1kB ± 0%   -5.02%  (p=0.000 n=10+10)
ParseModuleRulesBase/100-8             281kB ± 0%     266kB ± 0%   -5.17%  (p=0.000 n=10+10)
ParseModuleRulesBase/1000-8           2.80MB ± 0%    2.65MB ± 0%   -5.15%  (p=0.000 n=10+10)
ParseStatementBasicCall-8             5.27kB ± 0%    5.03kB ± 0%   -4.55%  (p=0.000 n=10+10)
ParseStatementMixedJSON-8             62.5kB ± 0%    60.5kB ± 0%   -3.15%  (p=0.000 n=10+10)
ParseStatementSimpleArray/1-8         5.12kB ± 0%    4.83kB ± 0%   -5.63%  (p=0.000 n=10+10)
ParseStatementSimpleArray/10-8        8.48kB ± 0%    8.19kB ± 0%   -3.40%  (p=0.000 n=10+10)
ParseStatementSimpleArray/100-8       46.4kB ± 0%    46.1kB ± 0%   -0.62%  (p=0.000 n=10+10)
ParseStatementSimpleArray/1000-8       413kB ± 0%     413kB ± 0%   -0.07%  (p=0.000 n=10+8)
ParseStatementNestedObjects/1x1-8     6.04kB ± 0%    5.70kB ± 0%   -5.56%  (p=0.000 n=10+10)
ParseStatementNestedObjects/5x1-8     9.17kB ± 0%    8.83kB ± 0%   -3.66%  (p=0.000 n=10+10)
ParseStatementNestedObjects/10x1-8    13.2kB ± 0%    12.8kB ± 0%   -2.55%  (p=0.000 n=10+10)
ParseStatementNestedObjects/1x5-8     11.6kB ± 0%    10.9kB ± 0%   -6.19%  (p=0.000 n=10+10)
ParseStatementNestedObjects/1x10-8    18.6kB ± 0%    17.4kB ± 0%   -6.45%  (p=0.000 n=10+10)
ParseStatementNestedObjects/5x5-8     3.81MB ± 0%    3.74MB ± 0%   -1.97%  (p=0.000 n=10+10)
ParseBasicABACModule-8                39.0kB ± 0%    37.2kB ± 0%   -4.67%  (p=0.000 n=10+8)

name                                old allocs/op  new allocs/op  delta
ParseModuleRulesBase/1-8                74.0 ± 0%      74.0 ± 0%     ~     (all equal)
ParseModuleRulesBase/10-8                549 ± 0%       549 ± 0%     ~     (all equal)
ParseModuleRulesBase/100-8             5.33k ± 0%     5.33k ± 0%     ~     (all equal)
ParseModuleRulesBase/1000-8            53.0k ± 0%     53.0k ± 0%     ~     (all equal)
ParseStatementBasicCall-8               51.0 ± 0%      51.0 ± 0%     ~     (all equal)
ParseStatementMixedJSON-8              1.02k ± 0%     1.02k ± 0%     ~     (all equal)
ParseStatementSimpleArray/1-8           39.0 ± 0%      39.0 ± 0%     ~     (all equal)
ParseStatementSimpleArray/10-8           115 ± 0%       115 ± 0%     ~     (all equal)
ParseStatementSimpleArray/100-8          839 ± 0%       839 ± 0%     ~     (all equal)
ParseStatementSimpleArray/1000-8       8.04k ± 0%     8.04k ± 0%     ~     (all equal)
ParseStatementNestedObjects/1x1-8       52.0 ± 0%      52.0 ± 0%     ~     (all equal)
ParseStatementNestedObjects/5x1-8        124 ± 0%       124 ± 0%     ~     (all equal)
ParseStatementNestedObjects/10x1-8       210 ± 0%       210 ± 0%     ~     (all equal)
ParseStatementNestedObjects/1x5-8        128 ± 0%       128 ± 0%     ~     (all equal)
ParseStatementNestedObjects/1x10-8       223 ± 0%       223 ± 0%     ~     (all equal)
ParseStatementNestedObjects/5x5-8      71.1k ± 0%     71.1k ± 0%     ~     (all equal)
ParseBasicABACModule-8                   697 ± 0%       697 ± 0%     ~     (all equal)
```

Signed-off-by: Patrick East <east.patrick@gmail.com>

patrick-east committed:
Add list of stuff to changelog

we should make it a little nicer before merging.. just getting a list
of everything that should probably be mentioned

Signed-off-by: Patrick East <east.patrick@gmail.com>

patrick-east committed:
parser_new.go -> parser.go

Signed-off-by: Patrick East <east.patrick@gmail.com>

patrick-east commented:
I think we're good to go now, probably just like squashing and polish up the change log entries however we want.
